everware/git_executor.py

```diff
@@ -0,0 +1,93 @@
+import git
+import re
+from tornado import gen
+from concurrent.futures import ThreadPoolExecutor
+
+
+class GitExecutor:
+
+    def __init__(self, repo_url, tmp_dir):
+        self._repo_url = repo_url
+        self._repo_dir = tmp_dir
+        self._repo_pointer = None
+        if '@' in repo_url:
+            self._processed_repo_url, self._repo_pointer = repo_url.split('@')
+        else:
+            parts = re.match(
+                r'(^.+?://[^/]+/[^/]+/.+?)(?:/|$)(tree|commit)?(/[^/]+)?',
+                repo_url
+            )
+            if not parts:
+                raise ValueError('Incorrect repo url')
+            self._processed_repo_url = parts.group(1)
+            if parts.group(3):
+                self._repo_pointer = parts.group(3)[1:]
+        if self._processed_repo_url.startswith('https') and\
+                self._processed_repo_url.endswith('.git'):
+            self._processed_repo_url = self._processed_repo_url[:-4]
+        if not self._repo_pointer:
+            self._repo_pointer = 'HEAD'
+
+    _git_executor = None
+
+    @property
+    def git_executor(self):
+        """single global git executor"""
+        cls = self.__class__
+        if cls._git_executor is None:
+            cls._git_executor = ThreadPoolExecutor(20)
+        return cls._git_executor
+
+    _git_client = None
+
+    @property
+    def git_client(self):
+        """single global git client instance"""
+        cls = self.__class__
+        if cls._git_client is None:
+            cls._git_client = git.Git()
+        return cls._git_client
+
+    def _git(self, method, *args, **kwargs):
+        """wrapper for calling git methods
+
+        to be passed to ThreadPoolExecutor
+        """
+        m = getattr(self.git_client, method)
+        return m(*args, **kwargs)
+
+    def git(self, method, *args, **kwargs):
+        """Call a git method in a background thread
+
+        returns a Future
+        """
+        return self.git_executor.submit(self._git, method, *args, **kwargs)
+
+    @gen.coroutine
+    def exec(self):
```

Comment:
Can we rearrange things so that we can have `executor.checkout_ref(ref)` instead? Then it would be clear what this method does.